java/code/src/com/redhat/rhn/domain/reactor/SaltEventFactory.java

```diff
@@ -43,26 +44,35 @@ private SaltEventFactory() {
 
     /**
      * Returns the approximate number of Salt events currently queued.
-     * @return the count
+     * @param threadCount the number of threads in the pool
+     * @return the list of events count per queue starting with a queue corresponding to events
+     *  without any minion ID. This queue is referred to as queue -1.
      */
-    public static long countSaltEvents() {
-        return (Long) singleton.lookupObjectByNamedQuery("SaltEvent.countSaltEvents", Collections.EMPTY_MAP);
+    @SuppressWarnings("unchecked")
+    public static List<Long> countSaltEvents(int threadCount) {
+        List<Object[]> countObjects = singleton.listObjectsByNamedQuery("SaltEvent.countSaltEvents",
+                    Collections.EMPTY_MAP);
+
+        List<Long> counts = new ArrayList<>(Collections.nCopies(threadCount + 1, 0L));
+        countObjects.forEach(count -> counts.set((int)count[0] + 1, (long)count[1]));
```

Comment:
I would suggest a more functional approach here (to avoid the `set` below): how about [reducing](https://docs.oracle.com/javase/8/docs/api/java/util/stream/Stream.html#reduce-T-java.util.function.BinaryOperator-) the `Stream` of `countObjects`, with a `Stream<Long>` as the `identity` parameter?

Reply:
that the query won't return the queues with no events, and we can't assume the queues counts are sorted. I don't really see how reduce would work there

Reply:
@cbosdo, my fault. Reducing in Java Streams has a narrower meaning than in other languages, so you are right, that won't work here. The closest solution I could find is here:

https://gist.github.com/moio/47620a66c6d633a129436d32ceeddb33

Arguably, the functional solution is more verbose, but it still feels good to me to have no side effects. I will not insist on this one, YMMV.

Reply:
Even if it's more verbose and basically does the same thing, I still find your way more readable

Reply:
:heavy_check_mark: